vbs.1
.TH WELCOME "1" "December 2003" "vbs" "User Commands"
.SH NAME
vbs \- simulate Verilog behavioral descriptions
.SH SYNOPSYS
\fBvbs\fR [\fIOPTION\fR]... \fISOURCE\fR...
.SH DESCRIPTION
Simulate SOURCE files.
.HP
\fB\-a\fR, \fB\-\-ascii_dump\fR
.IP
Dump using ascii format.
.HP
\fB\-c\fR, \fB\-\-compile\-only\fR
.IP
Stop after compilation, do not simulate.
.HP
\fB\-d\fR, \fB\-\-debug\-enable\fR=\fITYPE\fR
.IP
Enables the specified TYPE of debug output, where TYPE is one of the
following: sim_state, parser, symbol_table, time_wheel, event,
user[1-4], userx. This option may be used multiple times, but is only
available if debug is enabled.
.HP
\fB\-D\fR, \fB\-\-define\fR
.IP
Define macro for preprocessing (only for vpp).
.HP
\fB\-E\fR, \fB\-\-preprocess\-only\fR
.IP
Stop after preprocessing, do not compile.
.HP
\fB\-I\fR, \fB\-\-incdir\fR
.IP
Search directory for include files (only for vpp).
.HP
\fB\-l\fR, \fB\-\-debug\-outfile\fR=\fIFILE\fR
.IP
Specifies the file to write debug output. If not specified and debug
is enabled, output will go to stdout. Available only if debug is
enabled.
.HP
\fB\-q\fR, \fB\-\-quiet\fR
.IP
Do not output statistics.
.HP
\fB\-v\fR, \fB\-\-vcd_dump\fR
.IP
Dump using VCD format.
.HP
\fB\-x\fR, \fB\-\-delays\fR={min,typ,max}
.HP
\fB\-z\fR \fI#\fR
.IP
Compress dumpfile with zlib at the specified compression level, #
(from 1 to 9).
.HP
\fB\-h\fR, \fB\-\-help\fR
.IP
Display this message.
.SH "AUTHOR"
Written by Lay Hoon Tho and Jimen Ching.

This manual page was written by Shaun Jackman <sjackman@debian.org>
for the Debian system.
.SH "REPORTING BUGS"
Report bugs to Jimen Ching <jimen.ching@gmail.com>.
.SH "COPYRIGHT"
Copyright 1995, 1996, 1997, 2001, 2011 Lay Hoon Tho, Jimen Ching

This is free software; see the source for copying conditions. There
is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
.SH "SEE ALSO"
http://code.google.com/p/verilog-behavioral-simulator/


CHANGELOG
CHANGELOG

-------------------------------------------------------------------------------
2003-06-08  Jimen Ching  <jching@core2>

	* FAQ: Mention negative numbers are not supported.

2003-06-07  Jimen Ching  <jching@core2>

	* FAQ: Don't support older gcc.
	Other cleanup.

	* vbs.txt: Document added features.

	* src/vbs.l, src/vbs.y: Update copyright.

	* src/Makefile.in: Clean all configure generated files.

2003-06-06  Jimen Ching  <jching@core2>

	* src/configure, src/Makefile.in, src/configure.in:
	Fix build for Perl support.

	* EXAMPLES/ivl.sh: Additional test cases to ignore.

	* src/vbs.l, src/vbs.y: Fix parsing of numbers and optional delay.

	* src/p_expr.cc: Support parsing negative numbers.

2003-02-17  Jimen Ching  <jching@core2>

	* src/common/st_setup.cc, src/misc/decbase.h, src/moditm/misetup.cc, src/stmt/assign.cc, src/stmt/assign.h, src/stmt/case.cc, src/stmt/case.h, src/stmt/ifelse.cc, src/stmt/ifelse.h, src/stmt/loopstmt.cc, src/stmt/loopstmt.h, src/stmt/seqblk.cc, src/stmt/seqblk.h, src/stmt/stmtbase.h, src/stmt/taskenbl.cc, src/stmt/taskenbl.h:
	Remove unused code.

2003-02-03  Jimen Ching  <jching@core2>

	* BUGS: Update with missing feature.

	* src/common/event.h, src/common/time_whl.h:
	More (C++) standard conformance.

2003-01-21  Jimen Ching  <jching@core2>

	* src/expr/esetup.cc: Initialize variable to silence warning.

2003-01-20  Jimen Ching  <jching@core2>

	* EXAMPLES/ivl.sh, src/common/bvector.cc, src/common/bvector.h, src/expr/eeval.cc, src/expr/esetup.cc, src/moditm/d_setup.cc, src/stmt/ssetup.cc, src/stmt/strigger.cc:
	Many bug fixes.

	* src/moditm/misetup.cc:
	Do not move delay or event control when wrapping statements.

2002-12-26  Jimen Ching  <jching@core2>

	* EXAMPLES/ivl.sh:
	Fix detection of log file.  Cleanup shell command usage.

	* src/stmt/ssetup.cc, src/misc/msetup.cc, src/moditm/d_setup.cc, src/p_expr.cc:
	Detect unsupported constructs and report error.

	* src/common/bvector.h, src/expr/number.cc, src/common/bvector.cc:
	Add support for outputing strings.

	* src/expr/qstr.cc: Do not allocate more bits than necessary.

	* src/p_stmt.cc: Use error reporting API instead.

2002-12-21  Jimen Ching  <jching@core2>

	* src/expr/esetup.cc, src/expr/esetup.h, src/misc/lvalue.cc, src/misc/lvalue.h, src/misc/msetup.cc, src/misc/msetup.h:
	Specify size when setting up rval in assignment.

	* src/misc/decsetup.cc, src/misc/decsetup.h, src/stmt/ssetup.cc:
	Use correct statement when creating event cache.
	Specify size when setting up rval in assignment.

	* src/stmt/strigger.cc:
	Do not use parent when triggering pre/post assignment in for-loop.

	* src/moditm/misetup.cc:
	Schedule statement no matter whether there is a delay or not.

	* src/stmt/strigger.cc, EXAMPLES/contrib_inst.v, EXAMPLES/mi006_assign.v, EXAMPLES/stmt010_nbassign.v, EXAMPLES/wt_4_3p03p03p03p0.v, EXAMPLES/x_allev.v, EXAMPLES/x_counter.v, EXAMPLES/x_cport.v, EXAMPLES/x_dae.v, EXAMPLES/x_dump.v, EXAMPLES/x_gareth.v, EXAMPLES/x_inflp.v, EXAMPLES/x_monitor.v:
	$monitor/$strobe includes newline.

	* src/common/event.h, src/common/st_setup.cc, src/common/st_setup.h, src/common/st_task.cc, src/common/st_task.h, src/common/st_trig.cc, src/common/st_util.cc, src/expr/esetup.cc, src/expr/esetup.h, src/misc/decsetup.cc, src/misc/decsetup.h, src/misc/dectrig.cc, src/misc/dectrig.h, src/misc/evntexpr.cc, src/misc/evntexpr.h, src/misc/msetup.cc, src/moditm/d_setup.cc, src/moditm/misetup.cc, src/sim.cc, src/stmt/assign.h, src/stmt/loopstmt.cc, src/stmt/loopstmt.h, src/stmt/seqblk.cc, src/stmt/seqblk.h, src/stmt/ssetup.cc, src/stmt/ssetup.h, src/stmt/stmtbase.h, src/stmt/strigger.cc, src/stmt/taskenbl.cc, src/stmt/taskenbl.h:
	Fix event delays in nested blocks.

2002-12-14  Jimen Ching  <jching@core2>

	* src/misc/decrdwr.cc: Removed unused event type.

2002-11-27  Jimen Ching  <jching@core2>

	* EXAMPLES/ivl.sh: Updated to latest IVL test suite.

	* src/common/bvector.h: Do not access out-of-range indices.

	* src/common/st_util.cc, src/expr/number.cc: Support %T format.

	* src/expr/qstr.cc: Support empty qouted strings.

	* src/vbs.y: Parse unsupported constructs and report it.

2002-11-22  Jimen Ching  <jching@core2>

	* EXAMPLES/ivl.sh: Report why test cases are inconclusive.

	* src/vbs.y: Specify unsupported constructs.

	* src/common/st_util.cc, src/expr/esetup.cc, src/stmt/ssetup.cc:
	Scope list should only modify the module instance scope when setting
	up the function/task declarations.

	* src/vbs.y:
	Issue error or parser will not detect unsupported constructs.

	* src/common/event.h, src/vbs.y: Fix compile errors with g++ 3.2.

2002-11-21  Jimen Ching  <jching@core2>

	* EXAMPLES/ivl.sh: Handle erroneous test cases.

	* src/common/st_node.h, src/common/st_util.cc, src/expr/esetup.cc, src/moditm/d_setup.cc, src/moditm/misetup.cc, src/stmt/ssetup.cc, src/sim.cc:
	Use instance scope in hierarchical reference.

	* src/common/st_setup.cc: Correctly order the scopes.

	* src/common/dstack.h: Actually behave like a stack.

2002-11-16  Jimen Ching  <jching@core2>

	* src/common/sym_tab.cc, src/common/sym_tab.h:
	Do not waste time making a copy of the scope list.

	* src/common/tmpl_utl.h, src/stmt/strigger.cc:
	Put reference of list type in template parameter.

2002-11-14  Jimen Ching  <jching@core2>

	* src/common/bvector.cc:
	Clear result variable, because it is used in calculation algorithm.

	* src/common/hash.cc, src/common/hash.h, src/common/st_setup.cc, src/common/st_util.cc, src/common/sym_tab.cc, src/common/sym_tab.h, src/moditm/misetup.cc, src/sim.cc:
	Allow module instance name to be the same as module definition name.
	Fix scope lookup problem when this is allowed.

	* src/moditm/d_setup.cc: Code cleanup.
	Should only lookup net symbol in local scope when declaring it.

	* src/expr/esetup.cc, src/stmt/ssetup.cc: Code cleanup.

2002-11-02  Jimen Ching  <jching@core2>

	* src/common/st_setup.cc, src/moditm/misetup.cc, src/sim.cc:

src/vbs.cc
// Verilog Behavioral Simulator
// Copyright (C) 2011 Jimen Ching
//
// This file is part of the Verilog Behavioral Simulator package.
// See the file COPYRIGHT for copyright and disclaimer information.
// See the file COPYING for the licensing terms and conditions.
// See the file CONTRIBUTORS for a list of contributing authors.
//
// vbs.cc

#include "common/error.h"
#include "common/dumpinfo.h"
#include "common/scp_tab.h"
#include "common/sym_tab.h"
#include "common/event.h"
#include "common/time_whl.h"
#include "expr/emon.h"
#include "misc/msetup.h"
#include "stmt/stmtbase.h"
#include "stmt/ssetup.h"
#include "moditm/misetup.h"
#include "moditm/d_setup.h"
#include "vbs.h"


vbs_engine::vbs_engine()
	{
	_dumpinfo = new dump_info;
	_scopetable = new scope_table;

	_symboltable = new symbol_table;
	_eventqueue = new event_queue<stmt_base>;
	_timewheel = new time_wheel<stmt_base>;

	_modulelist = new std::list<hash_value>;
	}

vbs_engine::~vbs_engine()
	{
	delete _dumpinfo;
	delete _scopetable;

	// The rest needs to be deleted in a specific order to allow any
	// smart pointers to deallocate first.
	delete _modulelist;

	delete _timewheel;
	delete _eventqueue;
	delete _symboltable; // Delete last since it contains everything.
	}

void
vbs_engine::initialize()
	{
	if (_instance == 0)
		_instance = new vbs_engine;

	// Clear other global variables to restart the engine.

	// Reset parser error code reporting.
	extern vbs_error::value_type parse_error_code;
	parse_error_code = vbs_error::SE_NONE;

	// Reallocate event list used by monitor_expr.
	monitor_expr::reset();

	// Reallocate postponed list of postponed setup statements.
	setup_stmt::reset();

	// Reallocate statements for port IO.
	setup_port::reset();

	// Reallocate statement list.
	setup_module_item::reset();

	// Reallocate statement list.
	decl_setup::reset();
	}

void
vbs_engine::reset()
	{
	delete _instance;
	_instance = 0;

	// Reallocate instance and perform other initialization.
	initialize();
	}

vbs_engine *vbs_engine::_instance;
